File: src/mode.cpp
#include <iostream>
#include <ros/ros.h>
#include <kkctbn2019/Mode.h>
#include <mavros_msgs/RCIn.h>
#include <sstream>
ros::Publisher mode_publisher;


void pwmModeCallback(const mavros_msgs::RCIn::ConstPtr& msg) {
    ROS_WARN("%d", sizeof(msg));    
    int pwm = msg->channels[7];
    
    /**  
    if (pwm < 1400) {
        kkctbn2019::Mode mode;
        mode.value = kkctbn2019::Mode::HOLD;
        mode_publisher.publish(mode);
    } else if (pwm > 1600) {
        kkctbn2019::Mode mode;
        mode.value = kkctbn2019::Mode::MANUAL;
        mode_publisher.publish(mode);
    } else {
        kkctbn2019::Mode mode;
        mode.value = kkctbn2019::Mode::AUTO;
        mode_publisher.publish(mode);
    }
    */
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "mode");
    ros::NodeHandle nh;

    ros::Subscriber pwm_mode_subscriber = nh.subscribe<mavros_msgs::RCIn>("/mavros/rc/in", 8, pwmModeCallback);

    mode_publisher = nh.advertise<kkctbn2019::Mode>("/makarax/mode", 8);

    ROS_WARN("mode is active");

    ros::spin();
    
    return 0;
}

File: scripts/mode.py
#!/usr/bin/env python
import rospy
from kkctbn2019.msg import Mode
from mavros_msgs.msg import RCIn

def pwm_mode_callback(msg):
    try:
        pwm = msg.channels[7]
    except IndexError:
        pwm = 1700
    if (pwm < 1400):
        mode = Mode()
        mode.value = Mode.HOLD;
        mode_publisher.publish(mode);
    elif (pwm > 1600):
        mode = Mode()
        mode.value = Mode.ARMED;
        mode_publisher.publish(mode);

if __name__ == '__main__':
    rospy.init_node("mode")
    rc_subscriber = rospy.Subscriber("/mavros/rc/in", RCIn, pwm_mode_callback)
    mode_publisher = rospy.Publisher("/makarax/mode", Mode, queue_size=8);

    rospy.spin()



